File: typescript/src/sos/SosBroadcastService.ts
/**
 * Default SOS service. Originates and re-floods SOS broadcasts.
 *
 * Dedups by packet ID; rate-limited to MAX_SOS_BROADCASTS_PER_HOUR originations
 * per rolling hour.
 *
 * SPDX-License-Identifier: MIT
 */

import {
  MAX_SOS_BROADCASTS_PER_HOUR,
  SOS_PRIORITY,
  SOS_TTL,
} from "../constants.js";
import {
  BackendClient,
  IncentiveProvider,
  NoopBackendClient,
  NoopIncentiveProvider,
} from "../extensibility.js";
import { SosAcknowledgement, SosAlert } from "../models/index.js";
import { MeshPacket } from "../protocol/MeshPacket.js";
import { PacketType } from "../protocol/PacketType.js";
import { IMeshSender } from "../routing/IMeshSender.js";

export class SosBroadcastService {
  private readonly recentOrigins: Date[] = [];
  private readonly seen = new Set<string>();
  private readonly active = new Map<string, SosAlert>();

  onSosReceived?: (alert: SosAlert) => void;
  onSosResolved?: (broadcastId: string) => void;
  /**
   * Raised on the ORIGINATING node when a peer acknowledges receiving one of our active SOS
   * alerts — proof the emergency reached at least one device. Carries the responder and the
   * running distinct count. Mirrors the C# SosAcknowledged event.
   */
  onSosAcknowledged?: (ack: SosAcknowledgement) => void;

  constructor(
    private readonly sender: IMeshSender,
    private readonly backend: BackendClient = new NoopBackendClient(),
    private readonly incentives: IncentiveProvider = new NoopIncentiveProvider(),
  ) {}

  async broadcast(
    broadcastType: string,
    message: string | undefined,
    latitude: number,
    longitude: number,
    geohash?: string,
  ): Promise<boolean> {
    if (!broadcastType) throw new Error("broadcastType must not be empty");

    this.pruneOldOrigins();
    if (this.recentOrigins.length >= MAX_SOS_BROADCASTS_PER_HOUR) {
      return false;
    }
    this.recentOrigins.push(new Date());

    const alert: SosAlert = {
      id: crypto.randomUUID(),
      senderUhid: this.sender.localUhid,
      broadcastType,
      message,
      latitude,
      longitude,
      geohash,
      receivedAt: new Date(),
      acknowledgedBy: new Set<string>(),
    };
    this.active.set(alert.id, alert);

    const body = new TextEncoder().encode(
      JSON.stringify({
        broadcast_id: alert.id,
        broadcast_type: broadcastType,
        message: message ?? null,
        latitude,
        longitude,
        geohash: geohash ?? null,
      }),
    );

    const packet = new MeshPacket();
    packet.type = PacketType.SosBroadcast;
    packet.sourceUhid = this.sender.localUhid;
    packet.destinationUhid = "";
    packet.ttl = SOS_TTL;
    packet.priority = SOS_PRIORITY;
    packet.payload = body;
    this.seen.add(packet.id);

    await this.sender.broadcast(packet);
    await this.backend.syncSos(alert);
    return true;
  }

  resolve(broadcastId: string): void {
    if (this.active.delete(broadcastId)) {
      this.onSosResolved?.(broadcastId);
    }
  }

  getActiveAlerts(): SosAlert[] {
    return Array.from(this.active.values());
  }

  async handle(packet: MeshPacket): Promise<void> {
    if (packet.type !== PacketType.SosBroadcast) {
      throw new Error("expected PacketType.SosBroadcast");
    }
    if (this.seen.has(packet.id)) return;
    this.seen.add(packet.id);

    if (packet.sourceUhid === this.sender.localUhid) return;

    let data: {
      broadcast_id?: string;
      broadcast_type?: string;
      message?: string | null;
      latitude?: number;
      longitude?: number;
      geohash?: string | null;
    };
    try {
      data = JSON.parse(new TextDecoder().decode(packet.payload));
    } catch {
      return;
    }

    const alert: SosAlert = {
      id: data.broadcast_id ?? crypto.randomUUID(),
      senderUhid: packet.sourceUhid,
      broadcastType: data.broadcast_type ?? "sos",
      message: data.message ?? undefined,
      latitude: data.latitude ?? 0,
      longitude: data.longitude ?? 0,
      geohash: data.geohash ?? undefined,
      receivedAt: new Date(),
      acknowledgedBy: new Set<string>(),
    };
    this.active.set(alert.id, alert);
    this.onSosReceived?.(alert);

    // Acknowledge back to the originator so the sender learns their SOS reached a device.
    await this.sendSosAck(alert.id, packet.sourceUhid);

    if (packet.ttl > 1) {
      packet.ttl -= 1;
      await this.sender.broadcast(packet);
      await this.incentives.recordRelay(this.sender.localUhid, packet);
    }
  }

  /**
   * Pump an incoming SosAck packet into the service. On the ORIGINATING node it records the
   * responder against the matching active alert (deduping by responder UHID) and fires
   * onSosAcknowledged. No-op if the ack references an SOS this node did not originate.
   * Mirrors the C# HandleAckAsync.
   */
  async handleAck(packet: MeshPacket): Promise<void> {
    if (packet.type !== PacketType.SosAck) {
      throw new Error("expected PacketType.SosAck");
    }

    let data: { broadcast_id?: string; received_at_ms?: number };
    try {
      data = JSON.parse(new TextDecoder().decode(packet.payload));
    } catch {
      return;
    }
    if (!data.broadcast_id) return;

    // Only the ORIGINATOR holds this alert in `active`; every other node ignores the ack.
    const alert = this.active.get(data.broadcast_id);
    if (!alert) return;

    const responder = packet.sourceUhid;
    if (!responder) return;
    if (responder === this.sender.localUhid) return; // our own ack echoed back — ignore

    if (alert.acknowledgedBy.has(responder)) return; // already counted this responder — dedup
    alert.acknowledgedBy.add(responder);
    const total = alert.acknowledgedBy.size;

    this.onSosAcknowledged?.({
      broadcastId: data.broadcast_id,
      responderUhid: responder,
      totalAcknowledgements: total,
    });
  }

  // Send a directed SosAck back to the alert originator so the sender learns their emergency
  // reached this device. Best-effort: delivers when the originator is reachable as a next hop.
  private async sendSosAck(broadcastId: string, originatorUhid: string): Promise<void> {
    if (!originatorUhid) return;
    if (originatorUhid === this.sender.localUhid) return;

    const body = new TextEncoder().encode(
      JSON.stringify({
        broadcast_id: broadcastId,
        received_at_ms: Date.now(),
      }),
    );

    const ack = new MeshPacket();
    ack.type = PacketType.SosAck;
    ack.sourceUhid = this.sender.localUhid;
    ack.destinationUhid = originatorUhid;
    ack.ttl = SOS_TTL;
    ack.priority = SOS_PRIORITY;
    ack.payload = body;

    await this.sender.send(ack, originatorUhid);
  }

  private pruneOldOrigins(): void {
    const cutoff = Date.now() - 3600_000;
    while (this.recentOrigins.length > 0 && this.recentOrigins[0]!.getTime() < cutoff) {
      this.recentOrigins.shift();
    }
  }
}

File: swift/Sources/AetherNetProtocol/SOS/SosBroadcastService.swift
// SPDX-License-Identifier: MIT

import Foundation

/// SOS broadcast service. Originates and re-floods SOS broadcasts.
/// Dedups by packet ID; rate-limited to MAX_SOS_BROADCASTS_PER_HOUR per rolling hour.
public actor SosBroadcastService {
    private let sender: any MeshSender
    private let backend: any BackendClient
    private let incentives: any IncentiveProvider

    private var recentOrigins: [Date] = []
    private var seen: Set<UUID> = []
    private var activeAlerts: [UUID: SosAlert] = [:]

    public var onSosReceived: (@Sendable (SosAlert) -> Void)?
    public var onSosResolved: (@Sendable (UUID) -> Void)?
    /// Raised on the ORIGINATING node when a peer acknowledges receiving one of our active SOS
    /// alerts — proof the emergency reached at least one device. Carries the responder and the
    /// running distinct count.
    public var onSosAcknowledged: (@Sendable (SosAcknowledgement) -> Void)?

    public init(
        sender: any MeshSender,
        backend: any BackendClient = NoopBackendClient(),
        incentives: any IncentiveProvider = NoopIncentiveProvider()
    ) {
        self.sender = sender
        self.backend = backend
        self.incentives = incentives
    }

    public func setOnSosReceived(_ callback: (@Sendable (SosAlert) -> Void)?) {
        onSosReceived = callback
    }

    public func setOnSosResolved(_ callback: (@Sendable (UUID) -> Void)?) {
        onSosResolved = callback
    }

    public func setOnSosAcknowledged(_ callback: (@Sendable (SosAcknowledgement) -> Void)?) {
        onSosAcknowledged = callback
    }

    public func broadcast(
        broadcastType: String,
        message: String?,
        latitude: Double,
        longitude: Double,
        geohash: String? = nil
    ) async -> Bool {
        guard !broadcastType.isEmpty else { return false }
        pruneOldOrigins()
        if recentOrigins.count >= ProtocolConstants.maxSosBroadcastsPerHour { return false }
        recentOrigins.append(Date())

        let alert = SosAlert(
            senderUhid: sender.localUhid,
            broadcastType: broadcastType,
            message: message,
            latitude: latitude,
            longitude: longitude,
            geohash: geohash
        )
        activeAlerts[alert.id] = alert

        let body = encodeSosWire(
            broadcastId: alert.id,
            broadcastType: broadcastType,
            message: message,
            latitude: latitude,
            longitude: longitude,
            geohash: geohash
        )

        var packet = MeshPacket(
            type: .sosBroadcast,
            sourceUhid: sender.localUhid,
            destinationUhid: "",
            ttl: ProtocolConstants.sosTtl,
            priority: ProtocolConstants.sosPriority,
            payload: body
        )
        packet.timestampMs = Int64(Date().timeIntervalSince1970 * 1000)
        seen.insert(packet.id)

        _ = await sender.broadcast(packet)
        _ = await backend.syncSos(alert)
        return true
    }

    public func resolve(_ broadcastId: UUID) {
        if activeAlerts.removeValue(forKey: broadcastId) != nil {
            onSosResolved?(broadcastId)
        }
    }

    public func getActiveAlerts() -> [SosAlert] {
        Array(activeAlerts.values)
    }

    public func handle(_ packet: MeshPacket) async {
        guard packet.type == .sosBroadcast else { return }
        if !seen.insert(packet.id).inserted { return }
        if packet.sourceUhid == sender.localUhid { return }

        let parsed = parseSosWire(packet.payload)
        let alert = SosAlert(
            id: parsed?.broadcastId ?? UUID(),
            senderUhid: packet.sourceUhid,
            broadcastType: parsed?.broadcastType ?? "sos",
            message: parsed?.message,
            latitude: parsed?.latitude ?? 0,
            longitude: parsed?.longitude ?? 0,
            geohash: parsed?.geohash,
            receivedAt: Date()
        )
        activeAlerts[alert.id] = alert
        onSosReceived?(alert)

        // Acknowledge back to the originator so the sender learns their SOS reached a device.
        await sendSosAck(broadcastId: alert.id, to: packet.sourceUhid)

        if packet.ttl > 1 {
            var fwd = packet
            fwd.ttl = packet.ttl - 1
            _ = await sender.broadcast(fwd)
            await incentives.recordRelay(localUhid: sender.localUhid, packet: fwd)
        }
    }

    /// Pump an incoming ``PacketType/sosAck`` packet into the service. On the originating node it
    /// records the responder against the matching active alert (deduping by responder UHID) and
    /// fires ``onSosAcknowledged``. No-op if the ack references an SOS this node did not originate
    /// (only the originator holds it in `activeAlerts`), or one it has already resolved.
    public func handleAck(_ packet: MeshPacket) async throws {
        guard packet.type == .sosAck else {
            throw SosError.unexpectedPacketType(expected: .sosAck, actual: packet.type)
        }

        guard let body = parseSosAckWire(packet.payload) else { return }

        // Only the ORIGINATOR holds this alert in activeAlerts; every other node ignores the ack.
        guard var alert = activeAlerts[body.broadcastId] else { return }

        let responder = packet.sourceUhid
        if responder.isEmpty { return }
        if responder == sender.localUhid { return } // our own ack echoed back — ignore

        // Dedup by responder UHID (the packet source, NOT the payload).
        guard alert.acknowledgedBy.insert(responder).inserted else { return }
        activeAlerts[body.broadcastId] = alert
        let total = alert.acknowledgedBy.count

        onSosAcknowledged?(SosAcknowledgement(
            broadcastId: body.broadcastId,
            responderUhid: responder,
            totalAcknowledgements: total
        ))
    }

    /// Send a directed ``PacketType/sosAck`` back to the alert originator so the sender learns their
    /// emergency reached this device. Best-effort: delivers when the originator is reachable as a
    /// next hop. Skips empty or self originators.
    private func sendSosAck(broadcastId: UUID, to originatorUhid: String) async {
        if originatorUhid.isEmpty { return }
        if originatorUhid == sender.localUhid { return }

        let body = encodeSosAckWire(
            broadcastId: broadcastId,
            receivedAtMs: Int64(Date().timeIntervalSince1970 * 1000)
        )

        let ack = MeshPacket(
            type: .sosAck,
            sourceUhid: sender.localUhid,
            destinationUhid: originatorUhid,
            ttl: ProtocolConstants.sosTtl,
            priority: ProtocolConstants.sosPriority,
            payload: body
        )

        _ = await sender.send(ack, nextHopUhid: originatorUhid)
    }

    private func pruneOldOrigins() {
        let cutoff = Date().addingTimeInterval(-3600)
        recentOrigins.removeAll { $0 < cutoff }
    }
}

// ─── snake_case JSON wire ───

private struct SosWire: Codable {
    @LowercaseUUIDCoding var broadcast_id: UUID
    let broadcast_type: String
    let message: String?
    let latitude: Double
    let longitude: Double
    let geohash: String?
}

private func encodeSosWire(
    broadcastId: UUID,
    broadcastType: String,
    message: String?,
    latitude: Double,
    longitude: Double,
    geohash: String?
) -> Data {
    let w = SosWire(
        broadcast_id: broadcastId,
        broadcast_type: broadcastType,
        message: message,
        latitude: latitude,
        longitude: longitude,
        geohash: geohash
    )
    return (try? JSONEncoder().encode(w)) ?? Data()
}

private func parseSosWire(
    _ data: Data
) -> (broadcastId: UUID, broadcastType: String, message: String?, latitude: Double, longitude: Double, geohash: String?)? {
    guard let w = try? JSONDecoder().decode(SosWire.self, from: data) else { return nil }
    return (w.broadcast_id, w.broadcast_type, w.message, w.latitude, w.longitude, w.geohash)
}

// ─── SosAck wire (PacketType 6) ───
//
// Serialises to snake_case keys, field order broadcast_id then received_at_ms, no whitespace,
// GUID lowercase-dashed. This is the byte-identity gate (fixtures/sos/vectors.json).

private struct SosAckWire: Codable {
    @LowercaseUUIDCoding var broadcast_id: UUID
    let received_at_ms: Int64
}

// JSONEncoder key order is non-deterministic (hash-seed-dependent per process); build by hand in
// field order so byte-identity holds. (parseSosAckWire uses JSONDecoder, which is order-independent.)
private func encodeSosAckWire(broadcastId: UUID, receivedAtMs: Int64) -> Data {
    Data("{\"broadcast_id\":\"\(broadcastId.uuidString.lowercased())\",\"received_at_ms\":\(receivedAtMs)}".utf8)
}

/// Test-only shim exposing the real ``SosAckWire`` serialization path (the struct itself stays
/// `private`) so byte-identity vectors in `fixtures/sos/vectors.json` can be verified.
internal func _sosAckWireBytesForTests(broadcastId: UUID, receivedAtMs: Int64) -> Data {
    encodeSosAckWire(broadcastId: broadcastId, receivedAtMs: receivedAtMs)
}

private func parseSosAckWire(_ data: Data) -> (broadcastId: UUID, receivedAtMs: Int64)? {
    guard let w = try? JSONDecoder().decode(SosAckWire.self, from: data) else { return nil }
    return (w.broadcast_id, w.received_at_ms)
}

/// Errors thrown by ``SosBroadcastService``.
public enum SosError: Error, Equatable {
    /// A packet handed to a typed handler had the wrong ``PacketType``.
    case unexpectedPacketType(expected: PacketType, actual: PacketType)
}
